File: scripts/project_structure.py
"""
Generate a JSON representation of the current project's directory structure.
Excludes common development-related directories and files like .git, node_modules,
cache directories, and build artifacts.
"""
import os
import json
from pathlib import Path

def should_ignore(name):
    """Check if a file or directory should be ignored."""
    ignore_patterns = {
        # Directories
        '.git', '__pycache__', 'node_modules', 'dist', 'build',
        'venv', 'env', '.env', '.venv', 'target',
        # Files
        '.DS_Store', '*.pyc', '*.pyo', '*.pyd', '.Python',
        '*.so', '*.egg', '*.egg-info', '*.coverage',
        # Cache directories
        '.pytest_cache', '.mypy_cache', '.ruff_cache', '.coverage',
        # IDE directories
        '.idea', '.vscode', '.vs', '*.swp', '*.swo'
    }

    return any(
        name == pattern or
        (pattern.startswith('*.') and name.endswith(pattern[1:])) or
        name.startswith('.')
        for pattern in ignore_patterns
    )

def scan_directory(path):
    """Recursively scan directory and return structure as dictionary."""
    if isinstance(path, str):
        path = Path(path)

    result = {}

    try:
        for item in path.iterdir():
            if should_ignore(item.name):
                continue

            if item.is_file():
                result[item.name] = None
            elif item.is_dir():
                subdir = scan_directory(item)
                if subdir:  # Only add non-empty directories
                    result[item.name] = subdir
    except PermissionError:
        return None

    return result

def generate_structure(directory='.'):
    """Generate project structure as JSON."""
    structure = {os.path.basename(os.path.abspath(directory)): scan_directory(directory)}
    return json.dumps(structure, indent=2)

if __name__ == '__main__':
    print(generate_structure())

File: @fiction/core/plugin-env/utils/projectStructure.ts
import type { FictionEnv } from '../index.js'
import type { ConfigFileGenerator } from './generate.js'

import fs from 'fs-extra'
import { log } from '../../plugin-log/index.js'

type DirectoryStructure = {
  [key: string]: DirectoryStructure | null
}

/**
 * Check if a file or directory should be ignored
 */
function shouldIgnore(name: string): boolean {
  const ignorePatterns = new Set([
    // Directories
    '.git',
    '__pycache__',
    'node_modules',
    'dist',
    'build',
    'venv',
    'env',
    '.env',
    '.venv',
    'target',
    // Files
    '.DS_Store',
    // Cache directories
    '.pytest_cache',
    '.mypy_cache',
    '.ruff_cache',
    '.coverage',
    // IDE directories
    '.idea',
    '.vscode',
    '.vs',
  ])

  const ignoreExtensions = new Set([
    '.pyc',
    '.pyo',
    '.pyd',
    '.so',
    '.egg',
    '.egg-info',
    '.coverage',
    '.swp',
    '.swo',
  ])

  return (
    ignorePatterns.has(name)
    || Array.from(ignoreExtensions).some(ext => name.endsWith(ext))
    || name.startsWith('.')
  )
}

/**
 * Recursively scan directory and return structure as object
 */
async function scanDirectory(dirPath: string): Promise<DirectoryStructure | null> {
  const result: DirectoryStructure = {}

  const { join } = await import('node:path')

  try {
    const items = await fs.readdir(dirPath, { withFileTypes: true })

    for (const item of items) {
      if (shouldIgnore(item.name)) {
        continue
      }

      const fullPath = join(dirPath, item.name)

      if (item.isFile()) {
        result[item.name] = null
      }
      else if (item.isDirectory()) {
        const subdir = await scanDirectory(fullPath)
        if (subdir && Object.keys(subdir).length > 0) {
          result[item.name] = subdir
        }
      }
    }

    return result
  }
  catch (error) {
    if ((error as NodeJS.ErrnoException).code === 'EACCES') {
      log.warn('scanDirectory', `Permission denied accessing ${dirPath}`)
      return null
    }
    throw error
  }
}

/**
 * Generate project structure and write to .fiction config directory
 */
export const generateProjectStructure: ConfigFileGenerator = async (args: {
  fictionEnv: FictionEnv
}): Promise<{ fileName: string, content: string }> => {
  const { fictionEnv } = args
  const context = 'generateProjectStructure'

  if (!fictionEnv.cwd) {
    throw new Error(`${context}: cwd not found`)
  }

  const structure = await scanDirectory(fictionEnv.monorepoRoot)

  return { fileName: 'projectStructure.json', content: JSON.stringify(structure, null, 2) }
}

/**
 * Generate project structure as string
 */
export async function getProjectStructure(directory = '.'): Promise<string> {
  const { resolve, basename } = await import('node:path')
  const absPath = resolve(directory)
  const structure = {
    [basename(absPath)]: await scanDirectory(absPath),
  }
  return JSON.stringify(structure, null, 2)
}
